### PR TITLE
OS-60:added user account verification via email

### PR DESCRIPTION
**Pull request by sargiss987**
when the user clicks on the register button with valid credentials user's data saves on DB, but the user is able to do nothing unless verifies his/her account via email. after successful verification user has already logged in. if verification token time expires(24 hours) and the user tries to verify with it(not success), he/she just gets another one.

### CODE REVIEW

File: os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java [hunk 1]
```diff
@@ -33,8 +54,30 @@ public User register(UserRegistrationDto userDto) {
...
+    User user =
+        userRepository.save(
+            UserRegistrationMapper.userRegistrationDtoToUser(userDto, passwordEncoder));
+    sendEmailToVerifyUserAccount(user);
+    return user;
+  }
+
+  private void sendEmailToVerifyUserAccount(User user) {
+    String token = UUID.randomUUID().toString() + user.getId();
+    long expiresAt = ZonedDateTime.now().plusDays(1).toInstant().toEpochMilli();
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:66]: the time zone of the client(browser) and backend server can be different

Comment (SevakMart) [reply os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:66]: also let's keep amount of time in properties

File: os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java [hunk 2]
```diff
@@ -33,8 +54,30 @@ public User register(UserRegistrationDto userDto) {
       throw new EmailAlreadyExistException();
     }
 
-    User user = UserRegistrationMapper.userRegistrationDtoToUser(userDto, passwordEncoder);
-    return userRepository.save(user);
+    User user =
+        userRepository.save(
+            UserRegistrationMapper.userRegistrationDtoToUser(userDto, passwordEncoder));
+    sendEmailToVerifyUserAccount(user);
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:60]: we should do it asynchronously take a look on @Async

Comment (SevakMart) [reply os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:60]: move the logic about email sending into the separate service CommunicationService with EmailService implementation

File: os-server/src/main/resources/templates/email/account-verification.html [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    <title>Email</title>
+</head>
+<body>
+<div style="color: black; font-style: normal; font-size: 14px;">
+    Hey <b th:text="${userName}"></b>
+    <p>Click the button blow to verify your account.</p>
+    <p> It is valid until <b th:text="${date}"></b>.</p>
+
+    <form th:method="post"
+          th:action="@{http://localhost:5000/api/v1/auth/register/verification}"
```
Comment (SevakMart) [inline os-server/src/main/resources/templates/email/account-verification.html:18]: the hostname will be different for different environments. so please keep this in as property variable

File: os-server/src/main/java/app/openschool/auth/AuthController.java [hunk 4]
```diff
@@ -76,6 +79,15 @@ public ResponseEntity<UserLoginDto> login(@RequestBody UserLoginRequest userLogi
     return new ResponseEntity<>(userLoginDto, jwtHeader, OK);
   }
 
+  @PostMapping("/register/verification")
+  public ResponseEntity<UserLoginDto> verifyAccount(
+      @ModelAttribute VerificationToken verificationToken) {
+    User user = authService.verifyAccount(verificationToken);
+    UserPrincipal userPrincipal = new UserPrincipal(user);
+    HttpHeaders jwtHeader = getJwtHeader(userPrincipal);
+    return new ResponseEntity<>(UserLoginMapper.toUserLoginDto(user), jwtHeader, OK);
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthController.java:88]: I guess response entity builder can be used instead

File: os-server/src/main/java/app/openschool/common/services/EmailSenderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    MimeMessage message = mailSender.createMimeMessage();
+    MimeMessageHelper helper = new MimeMessageHelper(message, "utf-8");
+    try {
+      helper.setSubject("Verify your account");
+      helper.setText(emailContent, true);
+      helper.setFrom("epam.open.school@gmail.com");
+      helper.setTo(toEmail);
+      mailSender.send(message);
+    } catch (MessagingException e) {
+      e.printStackTrace();
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/common/services/EmailSenderServiceImpl.java:29]: what if error?

Comment (SevakMart) [reply os-server/src/main/java/app/openschool/common/services/EmailSenderServiceImpl.java:29]: we may consider logging

File: os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java [hunk 6]
```diff
@@ -33,8 +54,30 @@ public User register(UserRegistrationDto userDto) {
...
+    String token = UUID.randomUUID().toString() + user.getId();
+    long expiresAt = ZonedDateTime.now().plusDays(1).toInstant().toEpochMilli();
+    String date = new SimpleDateFormat("dd MMM yyyy HH:mm:ss").format(new Date(expiresAt));
+    VerificationToken verificationToken = new VerificationToken(user, token, expiresAt);
+    verificationTokenRepository.save(verificationToken);
+    emailSender.sendEmail(
+        user.getEmail(), createEmailContent(user.getName(), date, verificationToken));
+  }
+
+  private String createEmailContent(
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:74]: move this to email service

**Review by SevakMart — CHANGES_REQUESTED** [review]
please fix comments

File: os-server/src/main/java/app/openschool/auth/AuthController.java [hunk 7]
```diff
@@ -73,7 +76,22 @@ public ResponseEntity<UserLoginDto> login(@RequestBody UserLoginRequest userLogi
...
+    User user = authService.verifyAccount(verificationToken);
+    UserPrincipal userPrincipal = new UserPrincipal(user);
+    HttpHeaders jwtHeader = getJwtHeader(userPrincipal);
+    return ResponseEntity.ok().headers(jwtHeader).body(UserLoginMapper.toUserLoginDto(user));
+  }
+
+  @GetMapping ("/{userId}/account/verification")
+  public ResponseEntity<Void> sendVerificationEmail(@PathVariable Long userId) {
+    authService.sendVerificationEmail(userId);
+    return ResponseEntity.ok(null);
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthController.java:94]: please try to build without passing null

File: os-server/src/main/java/app/openschool/auth/verification/VerificationToken.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Id
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  @Column(name = "id", nullable = false)
+  private Long id;
+
+  @Column(name = "token", nullable = false)
+  private String token;
+
+  @Column(name = "expires_data", nullable = false)
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/verification/VerificationToken.java:25]: didn't we agreed to store creation date ?

File: os-server/src/main/java/app/openschool/common/services/EmailSenderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  @Override
+  public void sendEmail(String toEmail, String emailContent, String subject) {
+    MimeMessage message = mailSender.createMimeMessage();
+    MimeMessageHelper helper = new MimeMessageHelper(message, "utf-8");
+    try {
+      helper.setSubject(subject);
+      helper.setText(emailContent, true);
+      helper.setFrom("epam.open.school@gmail.com");
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/common/services/EmailSenderServiceImpl.java:28]: move this into property files

File: os-server/src/main/java/app/openschool/common/services/EmailSenderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+      helper.setTo(toEmail);
+      mailSender.send(message);
+    } catch (MessagingException e) {
+      logger.error("Sending mail to " + toEmail + " failed");
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/common/services/EmailSenderServiceImpl.java:32]: you can use the following syntax: 
logger.error("Sending mail to {} failed", toEmail);

File: os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java [hunk 11]
```diff
@@ -51,6 +73,32 @@ public UserLoginDto login(String userEmail) {
     return UserLoginMapper.toUserLoginDto(userRepository.findUserByEmail(userEmail));
   }
 
+  @Override
+  public User verifyAccount(VerificationToken verificationToken) {
+    VerificationToken fetchedToken =
+        verificationTokenRepository.findVerificationTokenByToken(verificationToken.getToken());
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:79]: try to use Optional in such cases like this

**Review by SevakMart — CHANGES_REQUESTED** [review]
use Optional while returning from repositories

**Comment by SevakMart** [comment]
please remove log files from PR

File: os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java [hunk 12]
```diff
@@ -71,6 +87,33 @@ public UserLoginDto login(String userEmail) {
...
+        communicationService.sendEmailToVerifyUserAccount(fetchedToken.get().getUser());
+        throw new UserNotVerifiedException();
+      }
+    }
+    throw new UserNotVerifiedException();
+  }
+
+  @Override
+  public void sendVerificationEmail(Long userId) {
+    User user = userRepository.findUserById(userId);
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:110]: use Optional

File: os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java [hunk 13]
```diff
@@ -71,6 +87,33 @@ public UserLoginDto login(String userEmail) {
     return UserLoginMapper.toUserLoginDto(userRepository.findUserByEmail(userEmail));
   }
 
+  @Override
+  public User verifyAccount(VerificationToken verificationToken) {
+    Optional<VerificationToken> fetchedToken =
+        verificationTokenRepository.findVerificationTokenByToken(verificationToken.getToken());
+
+    if (fetchedToken.isPresent()) {
+      if (!verificationToken.isTokenExpired(fetchedToken.get().getCreatedAt(), expiresAt)) {
```
Comment (SevakMart) [inline os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:96]: why we can't check the expiration time without getting it from the database?

File: os-server/src/main/resources/application-local.yml [hunk 14]
```diff
@@ -5,4 +5,10 @@ spring:
     username: open-school-user
     password: open-school-password
   jpa:
-    show-sql: true
\ No newline at end of file
+    show-sql: true
+
+verification:
+  endpoint: 'http://localhost:5000/api/v1/auth/account/verification'
```
Comment (SevakMart) [inline os-server/src/main/resources/application-local.yml:11]: this host should not be hardcoded

File: os-server/src/main/resources/messages.properties [hunk 15]
```diff
@@ -7,10 +7,14 @@ exception.nonexistent.user.email.message=User with {0} email does not exist
...
 exception.bad.credentials.message=Invalid user email or password
+exception.unverified.user.message=The verification token is expired or not valid
 response.register.successful.message=You've successfully registered
 exception.not.valid.token=Invalid token
 exception.expired.token=Token is expired
 exception.not.matching.passwords=Passwords do not match
 password.sending.success.message=The verification code was sent to your email. Please check.
 password.reset.success.message=Your password has been successfully changed
+verification.success.message={0} you are verified
+
```
Comment (SevakMart) [inline os-server/src/main/resources/messages.properties:18]: remove redundant empty lines

File: os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java (hunk 13 above)
Comment (sargiss987) [reply os-server/src/main/java/app/openschool/auth/AuthServiceImpl.java:96]: In the isTokenExpired method, we get the difference between the token created date and the verification request date, then compare it with the expiration duration. I guess the database request is necessary to get the token's creation date. Or maybe we should save tokens in cache memory?

**Review by SevakMart — APPROVED** [review]

**Pull request merged**